src/transforms/geoip.rs

```diff
@@ -0,0 +1,116 @@
+extern crate maxminddb;
+
+use super::Transform;
+
+use crate::{
+    event::Event,
+    topology::config::{DataType, TransformConfig},
+};
+use serde::{Deserialize, Serialize};
+use string_cache::DefaultAtom as Atom;
+
+use std::net::IpAddr;
+use std::str::FromStr;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct GeoipConfig {
+    pub source: Atom,
+    pub database: String,
+    pub target: String,
+}
+
+pub struct Geoip {
+    pub source: Atom,
+    pub database: String,
+    pub target: String,
+}
+
+#[typetag::serde(name = "geoip")]
+impl TransformConfig for GeoipConfig {
+    fn build(&self) -> Result<Box<dyn Transform>, crate::Error> {
+        Ok(Box::new(Geoip::new(
+            self.source.clone(),
+            self.database.clone(),
+            self.target.clone(),
+        )))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+impl Geoip {
+    pub fn new(source: Atom, database: String, target: String) -> Self {
+        Geoip {
+            source: source,
+            database: database,
+            target: target,
+        }
+    }
+}
+
+impl Transform for Geoip {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        println!("Event: {:?}", event.as_log());
+        let ipaddress = event
+            .as_log()
+            .get(&self.source)
+            .map(|s| s.to_string_lossy());
+        if let Some(ipaddress) = &ipaddress {
+            let ip: IpAddr = FromStr::from_str(ipaddress).unwrap();
+            println!("Looking up {}", ip);
+            let reader =
+                maxminddb::Reader::open_readfile("/usr/local/share/GeoIP/GeoIP2-City.mmdb")
+                    .unwrap();
```

Comment:
Instead of opening the file every time we handle an event, we should open it once in the `build` method above and store it as a field on the `Geoip` struct. That way we only do the work to open and parse it one time, and we have a chance to return any errors to the user instead of just unwrapping.

Reply:
The mmap feature was originally implemented because caching issues had been observed when storing the mmdb files on NFS volumes. I've yet to encounter a scenario where this would be less performant (might cause jitter until all needed parts of the database are cached).

This has less impact if, as I would second, the database is set up once and then only ever queried.

How about allowing the path to the database to be specified via env var?

Reply:
Agreed, I think whether or not we're using the mmap (which is awesome and we should probably use!), we're better off just opening once and holding a handle.

> How about allowing the path to the database to be specified via env var?

For this case specifically, I think we'll want the file location to be a part of the transform's configuration. Users can set that via env var if they'd like, but we shouldn't have to worry about that at this point in the stack.